Feature: Add js-client step by step guide (#14)
aragon/osx-plugin-template-hardhat: packages/js-client/GUIDE.md

Using this template and the `@aragon/sdk-client-common` package, you will be able to build a JS client for your plugin in a few minutes.

The `@aragon/sdk-client-common` package provides a set tools to build a JS client for your plugin, it mainly gives you helpers and battle tested logic for handling connections with external services such as:

- Web3 providers
- Subgraph
- IPFS

This package handles for you the connection (fail prevention, error handling, connection retry...) and basic functionality of these services, so you can focus on building your plugin.

### Why a JS client?

As mentioned before, the JS client is a totally optional step in the process of creating an Aragon OS plugin. You can definitely build a plugin without a JS client, but you will be missing some benefits.

Aragon common packages will handle the connection to IPFS, Subgraph and blockchain providers, so you don't have to reinvent the wheel.

Additionally, is giving some structure to your plugin, if you plan on doing some complex logic in your plugin, it is a good idea to have a JS client to handle it and leave the UI handling only the UI logic.

## Getting started

> To get started with your client the first step is to create a nes JS/TS project you can use any tool you want.

Reviewer: ```suggestion
To get started with your client the first step is to create a nes JS/TS project you can use any tool you want.
```

> To get started with your client the first step is to create a nes JS/TS project you can use any tool you want.

Reviewer 2: ```suggestion
To get started with your client the first step is to create a new JS/TS project you can use any tool you want.
```

> Once you have your project set up yo need to install the `@aragon/sdk-client-common` package.

Reviewer 2: ```suggestion
Once you have your project set up you need to install the `@aragon/sdk-client-common` package.
```

- `MyPluginOverriddenState`: This type will be used to track what values have been modified and what values are default. This type should also extend the `OverriddenState` type from `@aragon/sdk-client-common`.

In this file you should also define the state type of your client, this should extend the `ClientState` type from `@aragon/sdk-client-common`. This type will be used to define the state of your client, this state will be stored in the context and it will be accessible by the modules.

```ts
// src/types.ts
import {ContextParams} from '@aragon/sdk-client-common';

export type MyPluginClientParams = ContextParams & {
  myCustomParam: string;
  myCustomBackendUrl: string;
  // Your custom params
};

export type MyPluginState = ClientState & {
  myCustomParam: string;
  // In the context you will be able to work with the params
  // so the state can be a client for a service instead of the
  // actual value
  myCustomBackendClient: MyCustomBackendClient[];
};

export type MyPluginOverriddenState = OverriddenState & {
  [key in keyof MyPluginContextState]: boolean;

> };

Reviewer 2: Im a little lost here. There isnt a good explanation of what these types are or how to define them. and when you look in the template its not obvious as to what they map to. especially the back end client

Author: I added some more explaining lines, please re review it

Reviewer 2: Awesome!
I have not followed the the guide and built a client yet but reading through it makes sense and im no longer lost  =]

```

### `src/context.ts`

> The context is the responsible for storing the state of the of the client and setting the default values for the params. The context is also responsible for doing any kind of operation such as modifying some parameter, or using a parameter for creating a client for a service.

Reviewer 2: what parameters are we talking about here (internal parameters), also what does it mean to use a parameter for creating a client for a service? is this another client other than the one this context is a part of?

Author: I updated this paragraph

This `Context` uses the previously defined `state` and `overriddenState` types.

```ts
// src/context.ts
export class MyPluginClientContext extends Context {
  protected state: MyPluginContextState;
  protected overridden: MyPluginOverriddenState;

  constructor(params?: Partial<MyPluginClientParams>) {
    super();

>     this.set(contextParams);

Reviewer 2: ```suggestion
    this.set(params);
```

  }

  public set(params: Partial<MyPluginClientParams>) {
    // First we need to set the values in the parent class
    super.set(params);
    // Then we can set the values in our context

>     if (params.myCustomParam) {

Reviewer 2: so in this set function we switch between all the things we added in `MyPluginState`

Author: Yep the idea is that you check that the param exists and then you set the overriden to true

Author: I added some comments here so it is clearer

      this.state.myCustomParam = params.myCustomParam;
      this.overridden.myCustomParam = true;
    }
    if (params.myCustomBackendUrl) {
      this.state.myCustomBackendClient = new MyCustomBackendClient(params.myCustomBackendUrl);
      this.overridden.myCustomBackendClient = true;
    }
  }
}
```

In the minimal version of this file you just receive the contextParams on the constructor, but you can also improve this, by receiving an aragon context and reusing it to create a context for your client.

```ts
// src/context.ts
export class MyPluginClientContext extends Context {
  protected state: MyPluginContextState;
  protected overridden: MyPluginOverriddenState;

  constructor(params?: Partial<MyPluginClientParams>, context?: Context) {
    super();
    if (context) {
      // copy the context properties to this
      Object.assign(this, context);
    }
    if (contextParams) {
      // override the context params with the ones passed to the constructor
      this.set(contextParams);
    }
  }

  // Other methods
}
```

> You can also add more methods and helpers in this calss that can make your life easieer whule building your client like `getters` or `setters` for the state or a function that sets the defualt values for the state.

Reviewer 2: ```suggestion
You can also add more methods and helpers in this class that can make your life easier while building your client like `getters` or `setters` for the state or a function that sets the defualt values for the state.
```

> You can also add more methods and helpers in this calss that can make your life easieer whule building your client like `getters` or `setters` for the state or a function that sets the defualt values for the state.

Reviewer 2: ```suggestion
You can also add more methods and helpers in this class that can make your life easier while building your client like `getters` or `setters` for the state or a function that sets the default values for the state.
```

This could be an example of a more complete Context class.

```ts
// src/context.ts
const DEFAULT_MY_CUSTOM_PARAM = 'default value';
const DEFAULT_MY_CUSTOM_BACKEND_URL = 'https://my-custom-backend.com';

export class MyPluginClientContext extends Context {
  protected state: MyPluginContextState;
  protected overridden: MyPluginOverriddenState;

  constructor(params?: Partial<MyPluginClientParams>, context?: Context) {
    super();
    if (context) {
      // copy the context properties to this
      Object.assign(this, context);
    }

>     if (contextParams) {

Reviewer 2: ```suggestion
    if (params) {
```

      // override the context params with the ones passed to the constructor

>       this.set(contextParams);

Reviewer 2: ```suggestion
      this.set(params);
```

      this.state.myCustomParam = DEFAULT_MY_CUSTOM_PARAM;
    }
    if (!this.overridden.myCustomBackendClient) {
      this.state.myCustomBackendClient = new MyCustomBackendClient(DEFAULT_MY_CUSTOM_BACKEND_URL);
    }
  }

  get myCustomParam() {
    return this.state.myCustomParam;
  }

  get myCustomBackendClient() {
    return this.state.myCustomBackendClient;
  }
}
```

### `src/client.ts`

The client class is the one that will be used to call functionality from the UI, it's main purpose is to initialize and expose the available modules and variables.

```ts
// src/client.ts
import {ClientCore} from '@aragon/sdk-client-common';

export class MyPluginClient extends ClientCore {
  public methods: MyCustomMethodsModule;
  public estimation: MyCustomEstimationModule;
  public myCustomParam: string;
  public myCustomBackendClient: MyCustomBackendClient;
  constructor(context: MyCustomContext) {
    super(context);
    this.methods = new MyCustomMethodsModule(context);
    this.estimation = new MyCustomEstimationModule(context);
    this.myCustomParam = context.myCustomParam;
    this.myCustomBackendClient = context.myCustomBackendClient;
  }
}
```

> Now when you use the client from an UI you will be able to access the modules and the variables.

Reviewer 2: ```suggestion
Now when you use the client from a UI you will be able to access the modules and the variables.
```

# Step-by-step guide to build a JS client

---

## Introduction

The main goal of this guide is to show how to build a JS client for your Aragon OSx plugin. This guide assumes that you have already created a plugin and that you have a basic understanding of how Aragon OS works.

An Aragon OSx plugin is composed of four mainly components, not all of them are mandatory:

- Contracts
- JS Client (optional)
- Subgraph (optional)
- UI (optional)

The JS client sits as a middle layer between the contracts and the UI. It is responsible for fetching data from the contracts and exposing it to the UI. It also handles the communication between the UI and the contracts.

Using this template and the `@aragon/sdk-client-common` package, you will be able to build a JS client for your plugin in a few minutes.

> The `@aragon/sdk-client-common` package provides a set tools to build a JS client for your plugin, it mainly gives you helpers and battle tested logic for handling connections with external services such as:

Reviewer 3: ```suggestion
The `@aragon/sdk-client-common` package provides a set tools to build a JS client for your plugin. It gives you helpers for handling connections with external services such as:
```

- Web3 providers
- Subgraph
- IPFS

This package handles for you the connection (fail prevention, error handling, connection retry...) and basic functionality of these services, so you can focus on building your plugin.

### Why a JS client?

> As mentioned before, the JS client is a totally optional step in the process of creating an Aragon OS plugin. You can definitely build a plugin without a JS client, but you will be missing some benefits.

Reviewer 3: ```suggestion
As mentioned before, the JS client is an optional step in the process of creating an Aragon OSx plugin. You can definitely build a plugin without a JS client, but you will be missing some benefits.
```

Aragon common packages will handle the connection to IPFS, Subgraph and blockchain providers, so you don't have to reinvent the wheel.

Additionally, is giving some structure to your plugin, if you plan on doing some complex logic in your plugin, it is a good idea to have a JS client to handle it and leave the UI handling only the UI logic.

## Getting started

To get started with your client the first step is to create a new JS/TS project you can use any tool you want.

Once you have your project set up you need to install the `@aragon/sdk-client-common` package.

```bash
npm install @aragon/sdk-client-common
// or
yarn add @aragon/sdk-client-common
```

Once you have the package installed you can start building your client.

> There are 3 main files that you will need to build your client:

Reviewer 3: Isn't this the wrong approach/writing style? The idea of this template is that everything is provided and already works. 
This guide sounds like you need to do all of that before you can start. But the thing is, that we already created it for the user.
